drivers/ethernet/eth_mcux.c

```diff
@@ -458,9 +496,23 @@ static void eth_rx(struct device *iface)
 		frame_length -= frag_len;
 	} while (frame_length > 0);
 
+#if defined(CONFIG_NET_VLAN)
+	{
+		struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+
+		if (ntohs(hdr->type) == NET_ETH_PTYPE_VLAN) {
+			struct net_eth_vlan_hdr *hdr_vlan =
+				(struct net_eth_vlan_hdr *)NET_ETH_HDR(pkt);
+
+			net_pkt_set_vlan_tci(pkt, ntohs(hdr_vlan->vlan.tci));
+			vlan_tag = net_pkt_vlan_tag(pkt);
+		}
+	}
+#endif
+
 	irq_unlock(imask);
 
-	if (net_recv_data(context->iface, pkt) < 0) {
+	if (net_recv_data(get_iface(context, vlan_tag), pkt) < 0) {
```

Comment:
Myeah, net_recv_data() is blocking us from doing this kind of relation-ship (vtag/iface) into ethernet L2 recv part. Too bad, because it looks not nice to do it in the driver directly (which then need to do some book-keeping vtag/iface) where ethernet L2 has already this info...

Maybe one solution: 
What about exposing an ethernet L2 function that would give you the iface if you provide the vtag? Such function could be called from drivers. So drivers would only need to get net_recv_data(net_eth_get_iface_by_vtab(iface, vtag), pkt) change?  At least that would prevent redundant place holder for the iface/vtag relationship, don't you think?

Reply:
This is a bit problematic. I actually originally had iface <-> vlan mapping in ethernet.c only but then moved it to driver. The reason was that driver needed to know something about network interface it is using and I did not want to traverse the iface<->vlan tag list again in ethernet.c. So the simplest solution was to let the driver to set the proper iface according to vlan tag in the driver.

Reply:
I don't see the problem here, driver is traversing a list of iface too. It's actually better to have this in ethernet.c: one code place and not scattering the same logic on all drivers. And the driver would still provide the main iface (the one that holds the dev pointer), so it would traverse only its virt ifaces.

Reply:
I will investigate this. It is couple of weeks ago that I needed to change the code so do not recall the exact issues I had.